core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1275,21 +1275,42 @@ int runTranslogRecovery(Engine engine, Translog.Snapshot snapshot) throws IOExce
         return opsRecovered;
     }
 
+    /** creates an empty index and translog and opens the engine **/
+    public void createIndexAndTranslog() throws IOException {
+        assert recoveryState.getRecoverySource().getType() == RecoverySource.Type.EMPTY_STORE;
+        assert shardRouting.primary() && shardRouting.isRelocationTarget() == false;
+        // note: these are set when recovering from the translog
+        final RecoveryState.Translog translogStats = recoveryState().getTranslog();
+        translogStats.totalOperations(0);
+        translogStats.totalOperationsOnStart(0);
+        globalCheckpointTracker.updateGlobalCheckpointOnReplica(SequenceNumbers.NO_OPS_PERFORMED, "index created");
+        innerOpenEngineAndTranslog(EngineConfig.OpenMode.CREATE_INDEX_AND_TRANSLOG, false);
+    }
+
+    /** opens the engine on top of the existing lucene engine but creates an empty translog **/
+    public void openIndexAndCreateTranslog(boolean forceNewHistoryUUID, long globalCheckpoint) throws IOException {
+        assert recoveryState.getRecoverySource().getType() != RecoverySource.Type.EMPTY_STORE &&
+            recoveryState.getRecoverySource().getType() != RecoverySource.Type.EXISTING_STORE;
+        SequenceNumbers.CommitInfo commitInfo = store.loadSeqNoInfo();
+        if (commitInfo.localCheckpoint < globalCheckpoint) {
+            throw new IllegalArgumentException(
+                "trying to create a shard whose local checkpoint [" + commitInfo.localCheckpoint + "] is > global checkpoint ["
+                    + globalCheckpoint + "]");
+        }
+        globalCheckpointTracker.updateGlobalCheckpointOnReplica(globalCheckpoint, "opening index with a new translog");
```

Comment:
In case of LOCAL_SHARDS/RESTORE, we could again call activatePrimary here.

Reply:
agreed. See other comment.